Always wrap errors

alexjeffburke opened the pull request:
This pull request addresses #1 - it ensures all errors that escape the library are httpErrors. This allows a really nice pattern where callers acting as HTTP handlers need only do:

function callback(err) {
    res.status(err.statusCode).send();
}

to pass the error code up. I discovered this when I got a TypeError because I'd passed undefined statusCode to status().

I'd appreciate an extra eyeball on the selected of the 'nested' errorMode in the assertion I added to the test suite.

Thanks, Alex J Burke.

alexjeffburke committed:
Bump unexpected version for better error matching.

Includes an unexpected-mitm bump due to API incompatibility.

papandreou commented:
Gut reaction: I'm not too fond of pretending that socket/TCP errors are HTTP errors and making up status codes. At least it should be detectable what the error originally was.

I like the idea of the error instances carrying around a suitable HTTP status code, though, as that just makes everything easier with no downside. How about creating a `socketerrors` package that has mappings for all the error codes (`ECONNABORTED`, `ECONNREFUSED`, `ECONNRESET`, `ENETUNREACH`, `ENETRESET`, `ENETDOWN`, `ETIMEDOUT` etc.) and an equivalent HTTP status code? (I guess most of the errors map to 502 or 504).

papandreou commented:
> I'd appreciate an extra eyeball on the selected of the 'nested' errorMode in the assertion I added to the test suite.

That looks exactly right. For golfing bonus points, you could turn the two assertions into a single `to call the callback with [no] error` one:

``` js
        .addAssertion('to call the callback with error', function (expect, subject, expectedError) {
            var flags = this.flags;
            this.errorMode = 'nested';
            return expect.promise(function (run) {
                subject(run(function (err) {
                    if (flags.no) {
                        if (err) {
                            throw err;
                        }
                    } else {
                        expect(err, 'to equal', expectedError);
                    }
                }));
            });
```

alexjeffburke commented:
No probs and I do see your point about other random socket errors becoming plain 500s.

But since you agree with the approach of always emitting httpErrors, I've got no problem getting to work on a socketerror module to do a decent mapping for all the various types (a la a certain other client we are both friendly with).

That would leave the case of some other exception, but I think in that case a 500 mapping would be accurate. I like this. I'll have a think about that assertion simplification too.

alexjeffburke commented:
Hey,

I've just pushed this again using the new socketerrors module. I like the way this reads now, but see if that looks better to your eye.

AJB

alexjeffburke committed:
Ensure all errors are httpErrors to avoid a dereferencing TypeError.

Try first to convert unknown errors using the socketerrors module. If this
does not succeed we return a generic internal server error containing the
message from within the original error.

papandreou commented:
This looks great. I came up with a feature req for the socketerrors module that could simplify one of these tests slightly: https://github.com/alexjeffburke/node-socketerrors/pull/1

Anyway, I'll merge this immediately.

papandreou commented:
Released with 3.3.0.
